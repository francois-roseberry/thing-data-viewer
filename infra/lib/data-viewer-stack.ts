import path from "path";

import { Duration, Stack } from "aws-cdk-lib";
import {
  Dashboard,
  GraphWidget,
  type IWidget,
  MathExpression,
} from "aws-cdk-lib/aws-cloudwatch";
import {
  Effect,
  type IPrincipal,
  PolicyStatement,
  Role,
  ServicePrincipal,
} from "aws-cdk-lib/aws-iam";
import { CfnTopicRule } from "aws-cdk-lib/aws-iot";
import { Runtime } from "aws-cdk-lib/aws-lambda";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import { type Construct } from "constructs";

export interface DataViewerStackProps {
  readonly region: string;
  readonly account: string;
}

const METRIC_NAMESPACE = "ThingData";
const METRIC_DIMENSION = "Device";

export class DataViewerStack extends Stack {
  constructor(scope: Construct, id: string, props: DataViewerStackProps) {
    super(scope, id);

    const { region, account } = props;

    const projectRoot = path.join(__dirname, "../../lambda/");
    const lambda = new NodejsFunction(this, "MqttToCloudwatchMetricsLambda", {
      runtime: Runtime.NODEJS_20_X,
      handler: "./src/index.handler",
      functionName: "MqttToCloudwatchMetricsLambda",
      projectRoot,
      entry: path.join(projectRoot, "src/index.js"),
      depsLockFilePath: path.join(projectRoot, "yarn.lock"),
      timeout: Duration.seconds(60),
      environment: {
        METRIC_NAMESPACE,
        METRIC_DIMENSION,
      },
      initialPolicy: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: [
            "logs:PutLogEvents",
            "logs:CreateLogStream",
            "logs:CreateLogGroup",
          ],
          resources: [`arn:aws:logs:${region}:${account}:*`],
        }),
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ["cloudwatch:PutMetricData"],
          resources: ["*"],
        }),
      ],
    });
    lambda.grantInvoke(new ServicePrincipal("iot.amazonaws.com"));

    // I don't seem to be able to put dimensions in the metric with this action. Might have to go with a lambda

    const ruleLoggingRole = new Role(this, "valmetal-putToS3Action-role", {
      assumedBy: new ServicePrincipal("iot.amazonaws.com") as IPrincipal,
    });

    // Allow action to send to S3
    const logPolicy = new PolicyStatement({
      effect: Effect.ALLOW,
      actions: [
        "logs:PutLogEvents",
        "logs:CreateLogStream",
        "logs:CreateLogGroup",
      ],
      resources: [`arn:aws:logs:${region}:${account}:*`],
    });
    ruleLoggingRole.addToPolicy(logPolicy);

    new CfnTopicRule(this, "ThingDataRule", {
      ruleName: "thing_data_rule",
      topicRulePayload: {
        awsIotSqlVersion: "2016-03-23",
        description: "Rule created to forward MQTT payloads to a lambda",
        sql: "SELECT *, timestamp() as timestamp, topic(1) as device from '$aws/rules/thing_data_rule/+'",
        actions: [
          {
            lambda: {
              functionArn: lambda.functionArn,
            },
          },
        ],
        errorAction: {
          cloudwatchLogs: {
            logGroupName: "rule-errors",
            roleArn: ruleLoggingRole.roleArn,
          },
        },
      },
    });

    // This expression will return a Set of 1 metric per Device (ex: 2 devices publish, then 2 metrics will be returned)
    // The metric label will be the device name.
    // When graphing this, it will then generate one data series per unique device
    const temperatureMetricSet = new MathExpression({
      expression: `SELECT AVG(Temperature) FROM SCHEMA(${METRIC_NAMESPACE}, ${METRIC_DIMENSION}) GROUP BY ${METRIC_DIMENSION}`,
      period: Duration.minutes(1),
      label: "",
    });

    const humidityMetricSet = new MathExpression({
      expression: `SELECT AVG(Humidity) FROM SCHEMA(${METRIC_NAMESPACE}, ${METRIC_DIMENSION}) GROUP BY ${METRIC_DIMENSION}`,
      period: Duration.minutes(1),
      label: "",
    });

    const dashboard = new Dashboard(this, "MqttMetricsDashboard", {
      dashboardName: "ThingDataViewer",
    });

    dashboard.addWidgets(
      new GraphWidget({
        left: [temperatureMetricSet],
        leftYAxis: {
          label: "Temperature (Â°C)",
          min: 10,
          max: 50,
          showUnits: false,
        },
        width: 12,
        height: 6,
        liveData: true,
      }) as IWidget,
      new GraphWidget({
        left: [humidityMetricSet],
        leftYAxis: {
          label: "Humidity (%)",
          min: 20,
          max: 70,
          showUnits: false,
        },
        width: 12,
        height: 6,
        liveData: true,
      }) as IWidget,
    );
  }
}
