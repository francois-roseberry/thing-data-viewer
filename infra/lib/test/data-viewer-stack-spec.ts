import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";

import { DataViewerStack } from "../data-viewer-stack";

describe("Data Viewer stack", () => {
  let stack: DataViewerStack;

  beforeEach(() => {
    const app = new App();
    stack = new DataViewerStack(app, "TestStack", {
      region: "region",
      account: "1234",
    });
  });

  it("should match snapshot", () => {
    expect(Template.fromStack(stack)).toMatchSnapshot();
  });
});
