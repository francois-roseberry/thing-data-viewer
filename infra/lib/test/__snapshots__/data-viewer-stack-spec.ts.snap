// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data Viewer stack should match snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "MqttMetricsDashboard04861D1E": {
      "Properties": {
        "DashboardBody": {
          "Fn::Join": [
            "",
            [
              "{"widgets":[{"type":"metric","width":12,"height":6,"x":0,"y":0,"properties":{"view":"timeSeries","region":"",
              {
                "Ref": "AWS::Region",
              },
              "","metrics":[[{"expression":"SELECT AVG(Temperature) FROM SCHEMA(ThingData, Device) GROUP BY Device","period":60}]],"yAxis":{"left":{"label":"Temperature (Â°C)","min":0,"max":100,"showUnits":false}},"liveData":true}},{"type":"metric","width":12,"height":6,"x":12,"y":0,"properties":{"view":"timeSeries","region":"",
              {
                "Ref": "AWS::Region",
              },
              "","metrics":[[{"expression":"SELECT AVG(Temperature) FROM SCHEMA(ThingData, Device) GROUP BY Device","period":60}]],"yAxis":{"left":{"label":"Humidity (%)","min":0,"max":100,"showUnits":false}},"liveData":true}}]}",
            ],
          ],
        },
        "DashboardName": "ThingDataViewer",
      },
      "Type": "AWS::CloudWatch::Dashboard",
    },
    "MqttToCloudwatchMetricsLambda34BF85AC": {
      "DependsOn": [
        "MqttToCloudwatchMetricsLambdaServiceRoleDefaultPolicyE8A3E50E",
        "MqttToCloudwatchMetricsLambdaServiceRole76A974D2",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "a4eb3312bbb5639b390f46b96fd8658c7e14ea9c64212fe214f54124c1226a4e.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "METRIC_DIMENSION": "Device",
            "METRIC_NAMESPACE": "ThingData",
          },
        },
        "FunctionName": "MqttToCloudwatchMetricsLambda",
        "Handler": "./src/index.handler",
        "Role": {
          "Fn::GetAtt": [
            "MqttToCloudwatchMetricsLambdaServiceRole76A974D2",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "MqttToCloudwatchMetricsLambdaInvokeX1kgdpeq4A1b3T4Qgx5teuCJ9kkxQXuTuwWc9FHExDE9E699F21": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "MqttToCloudwatchMetricsLambda34BF85AC",
            "Arn",
          ],
        },
        "Principal": "iot.amazonaws.com",
      },
      "Type": "AWS::Lambda::Permission",
    },
    "MqttToCloudwatchMetricsLambdaServiceRole76A974D2": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MqttToCloudwatchMetricsLambdaServiceRoleDefaultPolicyE8A3E50E": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutLogEvents",
                "logs:CreateLogStream",
                "logs:CreateLogGroup",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:logs:region:1234:*",
            },
            {
              "Action": "cloudwatch:PutMetricData",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "MqttToCloudwatchMetricsLambdaServiceRoleDefaultPolicyE8A3E50E",
        "Roles": [
          {
            "Ref": "MqttToCloudwatchMetricsLambdaServiceRole76A974D2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ThingDataRule": {
      "Properties": {
        "RuleName": "thing_data_rule",
        "TopicRulePayload": {
          "Actions": [
            {
              "Lambda": {
                "FunctionArn": {
                  "Fn::GetAtt": [
                    "MqttToCloudwatchMetricsLambda34BF85AC",
                    "Arn",
                  ],
                },
              },
            },
          ],
          "AwsIotSqlVersion": "2016-03-23",
          "Description": "Rule created to forward MQTT payloads to a lambda",
          "ErrorAction": {
            "CloudwatchLogs": {
              "LogGroupName": "rule-errors",
              "RoleArn": {
                "Fn::GetAtt": [
                  "valmetalputToS3Actionrole44C6EABD",
                  "Arn",
                ],
              },
            },
          },
          "Sql": "SELECT * from '$aws/rules/thing_data_rule'",
        },
      },
      "Type": "AWS::IoT::TopicRule",
    },
    "valmetalputToS3Actionrole44C6EABD": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "iot.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "valmetalputToS3ActionroleDefaultPolicy39757A7A": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutLogEvents",
                "logs:CreateLogStream",
                "logs:CreateLogGroup",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:logs:region:1234:*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "valmetalputToS3ActionroleDefaultPolicy39757A7A",
        "Roles": [
          {
            "Ref": "valmetalputToS3Actionrole44C6EABD",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
